[ML] Set default value of 30 days for model prune window

edsavage opened the pull request:
For new anomaly detection jobs, when the analysis config field `model_prune_window` is not set, use a default value of 30 days.

edsavage committed:
[ML] Set default value of 30 days for model prune window

For new jobs, when the analysis config field model_prune_window is not set, use a default value of 30 days.

droberts195 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -1379,8 +1379,7 @@ Advanced configuration option.
 low, important information may be removed from the model. Typically, set to
-`30d` or longer. If not set, model pruning only occurs if the model memory
-status reaches the soft limit or the hard limit.
+`30d` or longer. If not set a default value of 30 days is used.
```
```suggestion
`30d` or longer. If not set, model pruning only occurs if the model memory
 status reaches the soft limit or the hard limit. For jobs created in 8.1 and
 above, a default value of 30 days is set on job creation if no other value is
 set.
```

edsavage committed:
Update docs/reference/ml/ml-shared.asciidoc

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Address code review comments

Ensure that a default model_prune_window value is only ever set for new
jobs, never for those created from indexed configuration documents.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java:
```diff
@@ -133,7 +135,7 @@
     private final List<String> influencers;
     private final Boolean multivariateByFields;
-    private final TimeValue modelPruneWindow;
+    private TimeValue modelPruneWindow;
```
This class is intended to be immutable, and code that uses it will be assuming this.  So this field will have to be changed back to `final`.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:
```diff
@@ -1207,6 +1207,7 @@ public void validateModelSnapshotRetentionSettingsAndSetDefaults() {
                 && modelSnapshotRetentionDays > DEFAULT_DAILY_MODEL_SNAPSHOT_RETENTION_AFTER_DAYS) {
                 dailyModelSnapshotRetentionAfterDays = DEFAULT_DAILY_MODEL_SNAPSHOT_RETENTION_AFTER_DAYS;
             }
+            analysisConfig.setDefaultModelPruneWindowIfNoneProvided();
```
Since `analysisConfig` is immutable, this should be changed as follows:

1. Check if the current `analysisConfig` has a model prune window
2. If it doesn't, construct an `AnalysisConfig.Builder` from `analysisConfig`
3. Call `setModelPruneWindow` on that `AnalysisConfig.Builder`
4. Call `setAnalysisConfig` on the `Job.Builder` that this method relates to, passing the `AnalysisConfig.Builder` - this will replace the old `analysisConfig` on the `Job.Builder` with the adjusted one

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java:
```diff
@@ -298,6 +300,12 @@ public TimeValue getModelPruneWindow() {
         return modelPruneWindow;
     }
 
+    public void setDefaultModelPruneWindowIfNoneProvided() {
```
Remove this, as immutable classes shouldn't have setters.

droberts195 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -1380,7 +1380,9 @@ Affects the pruning of models that have not been updated for the given time
 low, important information may be removed from the model. Typically, set to
 `30d` or longer. If not set, model pruning only occurs if the model memory
-status reaches the soft limit or the hard limit.
+ status reaches the soft limit or the hard limit. For jobs created in 8.1 and
+ above, a default value of 30 days is set on job creation if no other value is
+ set.
```
It looks like this is indented by 1 space, probably because of the way my earlier suggestion went in directly via GitHub.

droberts195 reviewed (commented):
I made a few comments about maintaining immutability.

Also, the response right at the end of `docs/reference/ml/anomaly-detection/apis/put-job.asciidoc` needs the new field adding.

And there's a code formatting error - you need to run `./gradlew :x-pack:plugin:ml:spotlessApply` to fix it.

edsavage committed:
Attend to code review comments

Ensure to retain immutability of AnalysisConfig

edsavage committed:
Merge branch 'master' of github.com:elasticsearch/elasticsearch into default_model_prune_window

edsavage committed:
Suppress warnings on test case

droberts195 commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/JobManagerTests.java:
```diff
@@ -186,6 +186,7 @@ public void onFailure(Exception e) {
 
     // Check that for new jobs, a default value is set for the model_prune_window field,
     // if one is not present in the configuration.
+    @SuppressWarnings("unchecked")
```
Is it just `ActionListener<Boolean> listener = (ActionListener<Boolean>) invocation.getArguments()[2]` that needs this?  If so, you should be able to put the suppression immediately before that line instead of applying to the whole method.

edsavage replied:
No, there's a few others in that test method as well
```
/Users/eds/src/elasticsearch/es-new/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/JobManagerTests.java:206: warning: [unchecked] unchecked method invocation: method createJobResultIndex in class JobResultsProvider is applied to given types
        }).when(jobResultsProvider).createJobResultIndex(requestCaptor.capture(), any(ClusterState.class), any(ActionListener.class));
                                                        ^
  required: Job,ClusterState,ActionListener<Boolean>
  found:    Job,ClusterState,ActionListener
/Users/eds/src/elasticsearch/es-new/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/JobManagerTests.java:206: warning: [unchecked] unchecked conversion
        }).when(jobResultsProvider).createJobResultIndex(requestCaptor.capture(), any(ClusterState.class), any(ActionListener.class));
                                                                                                              ^
  required: ActionListener<Boolean>
  found:    ActionListener
error: warnings found and -Werror specified
1 error
2 warnings
```
Given this I think it's cleaner to apply the annotation to the whole method.

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -1380,7 +1380,9 @@ Affects the pruning of models that have not been updated for the given time
 `30d` or longer. If not set, model pruning only occurs if the model memory
-status reaches the soft limit or the hard limit.
+status reaches the soft limit or the hard limit. For jobs created in 8.1 and
+above, a default value of 30 days is set on job creation if no other value is
```
I think this can be simplified to:
```suggestion
later, the default value is `30d`.
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -1380,7 +1380,9 @@ Affects the pruning of models that have not been updated for the given time
-status reaches the soft limit or the hard limit.
+status reaches the soft limit or the hard limit. For jobs created in 8.1 and
+above, a default value of 30 days is set on job creation if no other value is
+set.
```
If you accept my other suggestion, this one is also required:
```suggestion

```

edsavage committed:
Update docs/reference/ml/ml-shared.asciidoc

Co-authored-by: Lisa Cawley <lcawley@elastic.co>

edsavage committed:
Update docs/reference/ml/ml-shared.asciidoc

Co-authored-by: Lisa Cawley <lcawley@elastic.co>

edsavage committed:
Skipping broken compatibility tests

Skipping tests for which setting a default model prune window of 30 days
is a breaking change to compatibility across versions

droberts195 commented on x-pack/plugin/build.gradle:
```diff
@@ -98,6 +98,26 @@ tasks.named("yamlRestTestV7CompatTransform").configure{ task ->
   task.skipTest("vectors/40_sparse_vector_special_cases/Vectors of different dimensions and data types", "not supported for compatibility")
   task.skipTest("vectors/50_vector_stats/Usage stats on vector fields", "not supported for compatibility")
   task.skipTest("roles/30_prohibited_role_query/Test use prohibited query inside role query", "put role request with a term lookup (deprecated) and type. Requires validation in REST layer")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given missing job", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given non-matching expression and allow_no_match", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given non-matching expression and not allow_no_match", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given top_n is 0", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given top_n is negative", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given default", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given top_n is 2", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given top_n is 3", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given top_n is greater than the job count", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given overall_score filter", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given exclude_interim", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given string start and end params", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given epoch start and end params", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given invalid start param", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given invalid end param", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given bucket_span", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given bucket_span and top_n is 2", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given bucket_span and overall_score filter", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given no matching jobs", "behaviour change #81377 - setting default value of 30 days for model prune window")
+  task.skipTest("ml/jobs_get_result_overall_buckets/Test overall buckets given bucket_span is smaller than max job bucket_span", "behaviour change #81377 - setting default value of 30 days for model prune window")
```
It's pretty bad for us to be losing all our overall buckets REST testing because of a new default, when this new default shouldn't impact the overall buckets functionality at all.

I think a better solution is to open a separate PR that explicitly sets the model prune window in the overall buckets REST tests.  That PR can then be backported to 8.0 and 7.16.  Once those backports are merged these skips can be removed.

Or if I've guessed wrongly about what the underlying problem is here and you don't think that would work, why does setting a default prune window break the overall buckets functionality?

edsavage replied:
Setting a default value of 30 days breaks these tests because they incorporate a job that has a bucket span of 17 minutes and we have a validation check that the `model_prune_window` value must be an integral multiple of the bucket span.

As you suggest the optimal solution is to backport a change to the tests that explicitly sets a compatible value for the `model_prune_window` then re-instate these tests.

droberts195 replied:
Oh, interesting.  I think this reveals a problem with the new default then.  It's not a good UX if someone submits a perfectly good job config and we add a default to it that's incompatible.

I think this PR should be changed to set the default to the closest multiple of the bucket span that's greater than or equal to 30 days.  Generally that will be 30 days, but for weird bucket spans it might be 30 days and 4 minutes or something.

edsavage replied:
Sure, although it may make messaging a little awkward ;-)

droberts195 replied:
> set the default to the closest multiple of the bucket span that's greater than or equal to 30 days

In fact, since we're introducing a formula we should future proof it in other ways too.  With a 1 week bucket span we'd be pruning models after 5 buckets.  I think we should keep models for at least 20 buckets, so:

```
model_prune_window = max(20 * bucketSpan, closest multiple of the bucket span that's greater than or equal to 30 days)
```

droberts195 replied:
> it may make messaging a little awkward

I think it's OK to message it as the greater of 30 days or 20 times bucket span.  The slight rounding for weird bucket spans can be an implementation detail, very unlikely to be seen by sensible users.

edsavage committed:
Merge branch 'master' into default_model_prune_window

edsavage committed:
Removing redundant test

The complexity of JobManager::putJob is such that it is now extremely
difficult to test its behaviour. Removing a test that endeavoured to
check that a creation time field was added to the job but ended up
actually testing nothing and always returned true.

Such checks will be better off done by integration tests.

edsavage committed:
Calculate model_prune_window in a robust manner

When calculating the default value for model_prune_window take into
account the existing value of bucket_span as the model_prune_window must
be an integral multiple of this.

Effectively the model_prune_window is now calculated as the greater of
20 * bucket_span and 30 days.

edsavage committed:
Merge branch 'master' of github.com:elasticsearch/elasticsearch into default_model_prune_window

edsavage committed:
Allow for model_prune_window expressed in hours

edsavage committed:
Fix failing yaml test

droberts195 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -1380,7 +1380,8 @@ Affects the pruning of models that have not been updated for the given time
 `30d` or longer. If not set, model pruning only occurs if the model memory
-status reaches the soft limit or the hard limit.
+status reaches the soft limit or the hard limit. For jobs created in 8.1 and
+later, the default value is the greater of `30d` or 20 * `bucket_span`.
```
@lcawl do you have a view on whether it should be "20 times `bucket_span`" instead of "20 * `bucket_span`"?

lcawl replied:
I'd vote for the former for clarity.

droberts195 reviewed (approved):
LGTM, but please wait for the docs question to be answered before merging

edsavage committed:
Addressed code review comments

Clarified documentation
